onnxruntime/core/framework/bfc_arena.cc

```diff
@@ -95,30 +87,54 @@ bool BFCArena::Extend(size_t rounded_bytes) {
     return new_mem;
   };
 
-  void* mem_addr = safe_alloc(bytes);
-
-  if (mem_addr == nullptr) {
-    static constexpr float kBackpedalFactor = 0.9f;
+  auto get_extend_bytes = [this, available_bytes](const size_t bytes) -> size_t {
+    size_t extend_bytes = 0;
+    if (arena_extend_strategy_ == ArenaExtendStrategy::kNextPowerOfTwo) {
+      // If curr_region_allocation_bytes_ is not enough to satisfy the
+      // allocation, keep multiplying by a power of two until that is
+      // sufficient.
+      bool increased_allocation = false;
+      while (bytes > curr_region_allocation_bytes_) {
+        curr_region_allocation_bytes_ *= 2;
+        increased_allocation = true;
+      }
 
-    // Try allocating less memory.
-    while (mem_addr == nullptr) {
-      bytes = RoundedBytes(static_cast<size_t>(bytes * kBackpedalFactor));
-      if (bytes < rounded_bytes)
-        break;
+      extend_bytes = std::min(curr_region_allocation_bytes_, available_bytes);
 
-      mem_addr = safe_alloc(bytes);
+      // we allocated the same number of bytes as the current region, so we have 2x that now
+      if (!increased_allocation) {
+        curr_region_allocation_bytes_ *= 2;
+      }
+    } else if (arena_extend_strategy_ == ArenaExtendStrategy::kSameAsRequested) {
+      // BFC Arena could cause internal and external fragmentation. But, running training with
+      // big batch size will be very sensitive to fragmentation. So, to avoid fragmentation,
+      // just extend arena with actual requested size.
+      extend_bytes = bytes;
```

Comment:
Curious as to whether you tested with kMaxInternalFragmentation set to a much much smaller amount. I found there was a lot of wasted memory with it set to 128MB and I couldn't see a consistent benefit from the value (whether a chunk gets split or not is dependent on the size of the requested allocation so the 128MB is applied inconsistently). 
        const int64_t kMaxInternalFragmentation = 128 << 20;  // 128mb

Reply:
@weixingzhang

---
In reply to: [396963919](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r396963919) [](ancestors = 396963919)

Reply:
For training, we are trying to push the memory boundary to run as big batch size as possible. Frankly, I didn't try small value and wasn't aware this setting. But for the issue which I was trying to solve is related to both internal and external fragmentation. So, I would say that even a small value for internal fragmentation is probably not OK for my case.

Reply:
I think the name of the setting is misleading as I'm not sure it does anything to prevent fragmentation. Seems to just waste memory, so it for training you're trying to use as much memory as possible it could hurt. Fuller explanation here: [https://github.com/microsoft/onnxruntime/pull/2766#issuecomment-570422251](https://github.com/microsoft/onnxruntime/pull/2766#issuecomment-570422251)

---
In reply to: [400585476](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r400585476) [](ancestors = 400585476)

Reply:
The new strategy is to let BFC allocate memory same as requested, then memory fragmentation won't happen and no memory will be wasted.

Reply:
adding Scott's response - Renaming it is fine with me. kMaxDeadBytesInChunk would possibly be more meaningful. If we don’t do the split the difference between the requested bytes (rounded_bytes) and chunk->size becomes inaccessible until the chunk is freed. 

// If we can break the size of the chunk into two reasonably large
// pieces, do so.  In any case don't waste more than
// kMaxInternalFragmentation bytes on padding this alloc.
const int64_t kMaxInternalFragmentation = 128 << 20;  // 128mb
if (chunk->size >= rounded_bytes * 2 ||
    static_cast<int64_t>(chunk->size) - rounded_bytes >= kMaxInternalFragmentation) {
  SplitChunk(h, rounded_bytes);
  chunk = ChunkFromHandle(h);  // Update chunk pointer in case it moved
}


---
In reply to: [405063748](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r405063748) [](ancestors = 405063748)